src/psyclone/tests/psyir/frontend/fparser2_save_stmts_test.py

```diff
@@ -0,0 +1,175 @@
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. R. Porter, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+''' Tests Fortran save statements in the fparser2 PSyIR front-end '''
+
+import pytest
+
+from psyclone.errors import GenerationError
+from psyclone.psyir.nodes import Routine
+from psyclone.psyir.symbols import (StaticInterface, DefaultModuleInterface,
+                                    AutomaticInterface, UnknownFortranType)
+
+
+def test_save_statement_module(fortran_reader):
+    '''Tests that save statement and attribute are correctly captured when
+    they appear in a module, irrespective of ordering.'''
+    code = '''
+      module my_mod
+        save :: var3
+        integer :: var1
+        integer, save :: var2
+        save :: var1
+        integer :: var3, var4
+      end module my_mod'''
+    psyir = fortran_reader.psyir_from_source(code)
+    symtab = psyir.children[0].symbol_table
+    var1 = symtab.lookup("var1")
+    assert var1.is_constant is False
+    assert isinstance(var1.interface, StaticInterface)
+    var2 = symtab.lookup("var2")
+    assert isinstance(var2.interface, StaticInterface)
+    var3 = symtab.lookup("var3")
+    assert isinstance(var3.interface, StaticInterface)
+    var4 = symtab.lookup("var4")
+    assert isinstance(var4.interface, DefaultModuleInterface)
+
+
+def test_default_save_module(fortran_reader):
+    '''
+    Test that all Symbols declared in a module are given a StaticInterface if
+    there is a SAVE statement.
+    '''
+    code = '''
+      module my_mod
+        integer :: var1, var4
+        integer :: var2
+        save
+        integer :: var3
+      end module my_mod'''
```

Comment:
I notice that a single declaration wiht "save" and "parameter" is not valid fortran. But what happens with this separated declaration statements, is:
```
        integer, parameter :: var1
        save
```
or
```
        integer, save :: var1
        parameter :: var 1
```
valid Fortran? Do we do correctly accept/decline them?

What about unsupported attributes? Do things like:
```
        integer, volatile :: var1
        save
```
are valid? and if they are, are we still leaving the interface as UnknownInterface?

Reply:
AFAICS, the standard doesn't forbid the first two cases and gfortran is happy with them. I've added a test using your suggested fragments.

Reply:
Your third example also appears to be valid. I've added a test for this case too.